src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -166,139 +244,109 @@ public static Response Apply(Response current, Stack<Request> requests, List<Res
-                if (call.Item1.Expression is QsExpressionKind<TypedExpression, Identifier, ResolvedType>.Identifier id &&
-                    id.Item1 is Identifier.GlobalCallable globalCallable)
-                {
-                    // Rebuild the method with the updated identifier
-                    ex = new TypedExpression(
-                        QsExpressionKind<TypedExpression, Identifier, ResolvedType>.NewCallLikeExpression(
-                            new TypedExpression(
-                                QsExpressionKind<TypedExpression, Identifier, ResolvedType>.NewIdentifier(
-                                    GetConcreteIdentifier(globalCallable, types),
-                                    id.Item2
-                                    ),
-                                call.Item1.TypeParameterResolutions,
-                                call.Item1.ResolvedType,
-                                call.Item1.InferredInformation,
-                                call.Item1.Range
-                                ),
-                            call.Item2
-                            ),
-                        ex.TypeParameterResolutions,
-                        ex.ResolvedType,
-                        ex.InferredInformation,
-                        ex.Range
-                        );
-                }
+                CurrentParamTypes.Add(kvp.Key, kvp.Value);
             }
 
-            return base.Transform(ex);
+            return ImmutableDictionary<QsTypeParameter, ResolvedType>.Empty;
         }
+    }
 
-        private Concretion CreateConcretion(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+    internal class ReplaceTypeParamCallsExpressionKind : ExpressionKindTransformation<ReplaceTypeParamCallsExpression>
+    {
+        private readonly GetConcreteIdentifierFunc GetConcreteIdentifier;
+        private Dictionary<QsTypeParameter, ResolvedType> CurrentParamTypes;
+
+        public ReplaceTypeParamCallsExpressionKind(ReplaceTypeParamCallsExpression expr,
+            Dictionary<QsTypeParameter, ResolvedType> currentParamTypes,
+            GetConcreteIdentifierFunc getConcreteIdentifier) : base(expr)
         {
-            Concretion result = new Concretion();
-            foreach (var kvp in typeDict)
-            {
-                result.Add((kvp.Key, kvp.Value));
-            }
-            return result;
+            GetConcreteIdentifier = getConcreteIdentifier;
+            CurrentParamTypes = currentParamTypes;
         }
 
-        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> onIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
         {
-            Concretion target = null;
-            if (types != null && !types.IsEmpty)
+            if (sym is Identifier.GlobalCallable global)
```

Comment:
It would be good to add an else case that throws an ArgumentException if sym is not a GlobalCallable but there are type argument specified (i.e. tArgs is not Null). If you do that, then that implies that this transformation will only succeed on valid code and fail if there was a compilation error - which is fine - at the benefit of that it is not a silent failure if we start supporting assigning type parameterized objects.

Reply:
Please check that I have handled this case appropriately.

Reply:
Looks good.